### PR TITLE
[Quant] Respect non_leaf_module_list for activation modules

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88498

Summary: This commit fixes the bug where `non_leaf_module_list`
was not respected for activation modules like `torch.nn.Sigmoid`
and `torch.nn.Tanh`. Today, these modules default to
`default_fixed_qparams_range_0to1_fake_quant`, and there is no
way to configure them to use any other activation_post_process
(e.g. FixedQParamsObserver) (see this [mapping](https://github.com/pytorch/pytorch/blob/dc00bb51b8d370bf3891f0edb2c6e0c2914e329a/torch/ao/quantization/quantization_mappings.py#L188-L193)).
`non_leaf_module_list` is a "list of non-leaf modules we want
to add observer" (see prepare docstring). If the user explicitly
specified to insert observers for these modules, we should respect
that instead of continuing to use the default.

Test Plan:
python test/test_quantization.py TestQuantizeEagerPTQStatic.test_activations_in_non_leaf_module_list

Reviewers: vkuzo, jerryzh168

Subscribers: vkuzo, jerryzh168

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

### CODE REVIEW

**Review by jerryzh168 — APPROVED** [review]
LGTM, I think we should also consider removing the hack and break the BC for these ops (I think we already did that in fx graph mode quant)

File: torch/ao/quantization/quantize.py [hunk 1]
```diff
@@ -214,12 +214,12 @@ def insert_activation_post_process(m, special_act_post_process=None):
             # activation_post_process are now added directly to nn.Sequentail/_FusedModule
             if needs_observation(child):
                 insert_activation_post_process(child)
-        elif _has_special_act_post_process(child):
-            special_act_post_process = _get_special_act_post_process(child)
-            insert_activation_post_process(child, special_act_post_process)
         elif non_leaf_module_list is not None and type_before_parametrizations(child) in non_leaf_module_list:
             if needs_observation(child):
                 insert_activation_post_process(child)
+        elif _has_special_act_post_process(child):
```
[the remark this answers is not in the document]
Comment (andrewor14) [reply torch/ao/quantization/quantize.py:220]: I think so. There's no validation logic for eager mode right now so we just override it here. Yes I think simply removing it will make things cleaner, though it'll be a broader scope change since it'll break a lot of existing use cases. I'll leave that for a future PR to unblock https://github.com/pytorch/pytorch/pull/88456.

**Comment by andrewor14** [comment]
Thanks, I'm merging this first to unblock https://github.com/pytorch/pytorch/pull/88456. We can discuss further if we want to do the validation for eager mode separately.

File: torch/ao/quantization/quantize.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewor14) [reply torch/ao/quantization/quantize.py:220]: Filed https://github.com/pytorch/pytorch/issues/88579
